Implemented smooth zoom

paul-solomon opened the pull request:
## Add Smooth Zoom Option for Continuous Zooming 🎯

Hey! I've been using this plugin and it's fantastic, really appreciate the work you've put into it! 

### The Issue
I noticed that the current zoom can feel a bit choppy since it jumps by discrete amounts on each key press. It's especially noticeable when you want to zoom quickly but smoothly (like when you have a broken mouse wheel 😅).

### What I Added
I've implemented an optional **Smooth Zoom** feature that lets you zoom continuously while holding the keys. 

### Changes Made
- Added a new `smoothZoom` config option (defaults to `false` so nothing changes for current users)
- When enabled, holding zoom keys creates smooth continuous zooming at 60 FPS
- The smooth zoom still respects your `zoomIncrement` setting - it just spreads it across multiple frames
- All the original functionality stays exactly the same when smooth zoom is off

### How It Works
- **Regular Mode**: Works exactly like before - press key, get instant zoom
- **Smooth Mode**: Hold key down for continuous smooth zooming (like having a working mouse wheel!)

### Why This is Cool
- Zero breaking changes - existing users won't even notice unless they want to try it
- Gives players the choice between quick discrete jumps or smooth scrolling
- Perfect for people with hardware issues
- Uses the same increment value, so it feels consistent with your existing settings

I've tested it pretty thoroughly and it feels great to use. Would love to hear your thoughts! Happy to make any adjustments you'd like.

Thanks again for such a useful plugin! Best regards, Paul.

paul-solomon committed:
- Add smoothZoom config option (default: false)
- Implement continuous zoom while holding keys
- Maintain backward compatibility with existing behavior
- Use configurable zoom increment for smooth zoom speed"

R4lfE commented on src/main/java/com/zoomkeybinding/ZoomKeybindingPlugin.java:
```diff
@@ -47,26 +59,141 @@ protected void startUp() throws Exception
+		}, 0, ZOOM_UPDATE_INTERVAL_MS, TimeUnit.MILLISECONDS);
+	}
+
+	private void stopZooming()
```
Call it 'stopSmoothZoom' to be consistent with the start function.

paul-solomon replied:
Resolved

R4lfE commented on src/main/java/com/zoomkeybinding/ZoomKeybindingPlugin.java:
```diff
@@ -47,26 +59,141 @@ protected void startUp() throws Exception
+
+	private void handleSmoothZoomKeyPressed(KeyEvent keyEvent)
+	{
+		if (keyEvent.getKeyCode() == config.zoomInKey().getKeyCode())
```
Collapse the if statements.

paul-solomon replied:
Resolved

R4lfE commented on src/main/java/com/zoomkeybinding/ZoomKeybindingPlugin.java:
```diff
@@ -47,26 +59,141 @@ protected void startUp() throws Exception
+	{
+		if (zoomTask != null && !zoomTask.isDone())
+		{
+			zoomTask.cancel(false);
```
Does this nullify the zoomTask? Maybe use an Optional to wrap the zoomTask to clarify intent. It should then be emptied here.

paul-solomon replied:
So this one was indeed not nullifying the zoomTask. I modified it so it does. I initially went with the suggested approach of wrapping it in an optional, but after a warning and a bit of research, it turns out that's a Java anti-pattern, so I just ended up nullifying instead. More info about the anti-pattern here: https://dzone.com/articles/optional-anti-patterns

R4lfE reviewed (commented):
Thanks for this nice addition! The smooth zooming works fine as is. I have some very slight remarks that you could take a look at.

As an optional remark that could potentially improve your feature, maybe add another config item for smooth zooming so the increments of smooth and discrete zooming are not linked at all. You could also try to scale the smooth zoom with the actual user's FPS.

Let me know when you think the PR is ready and I'll merge the changes.

paul-solomon committed:
Addressed PR comments

paul-solomon committed:
Implemented config option for separate smooth zoom value

paul-solomon commented:
@R4lfE Thanks a bunch for the feedback and I'm glad to hear you think this is a nice addition! I addressed the PR comments (left a comment about the Optional wrapping). I also went ahead and implemented the idea you had. I did not find a way to make the config options conditionally shown so I wrapped them in a section instead. I also didn't implement the FPS sync, seems it's a bit of a complex job and more trouble than it's worth. 

From my point of view this is ready to merge, please let me know if you need any additional changes. Once merged, how long does it usually take until the plugin gets updated in Runelite? Thanks!

R4lfE commented:
I'm not sure how long it will take for the plugin hub maintainers to update the commit hash. If I recall correctly, it didn't take too long. But like I said, I can't give you an estimate.
